### ISSUE TITLE
[thunderbird 68] is Virtual Identity compatible?

### ISSUE THREAD

**Issue by despecial** [absorb-it/Virtual-Identity#22]
TB 68 is out now and there is a big change for extensions.
Since a compatibility checker is missing, I would like to ask if this VI can be used with that version?

**Comment by kewlfft** [comment]
A conversion is necessary, these steps must be followed for conversion: https://developer.thunderbird.net/add-ons/updates

**Comment by bl00dy1837** [comment]
It doesn't work anymore.

**Comment by blueyed** [comment]
@absorb-it 
Are there plans to make it compatible?

**Comment by absorb-it** [comment]
Hi, currently busy I will have a look into required changes during the next months... Sorry, this will take a while.

Stay tuned, regards, Rene

**Comment by ferricoxide** [comment]
It's not until Thunderbird rolls out yet another extension-breaking update that you remember, "oh, yeah: this extension is one of the reasons why I still even use Thunderbird". 

=)

**Comment by blueyed** [comment]
@ferricoxide 
True.  It really became a must-have for me.
Luckily Thunderbird allows you to "Customize From Address..." manually.

**Comment by miskovics** [comment]
+1

I love this extension, please, make it compatible.

**Comment by ferricoxide** [comment]
Yeah, but that requires actual effort ....and I'm soooooooooper-lazy. :)

> @ferricoxide
> True. It really became a must-have for me.
> Luckily Thunderbird allows you to "Customize From Address..." manually.

**Comment by miskovics** [comment]
... and it is much safer and faster to do it with this add-on instead of the "Customize from Address..." function.

**Comment by absorb-it** [comment]
Hi,
after looking into the development path of thunderbird (see for instance [here](https://groups.google.com/forum/#!topic/tb-planning/-E8Yw6POxEE)) I can see that the only way will be to get the complete functionality into core thunderbird. This will take a while, sorry.

these are the steps:
1. add some own-server/catch-all option to accounts for composing messages with the right senders email ([bugzilla-link](https://bugzilla.mozilla.org/show_bug.cgi?id=1518025))
2. change from-field as free-editable for own-server/catch-all accounts
3. store mail address used for recipients together with collected addresses in address-book

Lets see how this goes on, but once this is done it will last way longer than changing some extension ever and ever again.

Regards, Rene

**Comment by mcanonic** [comment]
Too bad. Can we help you Rene in some way?
Thanks for your extension. 
I'm sure that you guys have already check if any other add-on could replace (in some way) this extension in the meanwhile.

**Comment by chris-fp** [comment]
As others have said - it's only when TB breaks VI that you remember it was there :)

It's absolutely vital that I have VI running so I've just downgraded back to v67 and won't be upgrading again until VI functionality is restored somehow - be that upgraded add-on or, even better, it's built into TB directly :D :D

**Comment by sukamenev** [comment]
Thanks for the extension! I also downgraded Thunderbird to 60.9.1.
Let me know if you need a donation.

**Comment by github-normalnutzer** [comment]
Hi there,

is there any information about progress at Rene to integrate the whole thing into the TB core?

Going to an older TB version to be compatible with VI doesn't seem to be the best way for me for such a long time, now almost 3 months.
I really miss the VI functionality.

If a donation could help, it would probably only need a call. :-)

**Comment by miskovics** [comment]
> I really miss the VI functionality.
> If a donation could help, it would probably only need a call. :-)

I'm also ready to donate.

**Comment by chris-fp** [comment]
Hi,

I'd also be happy to give a donation to get this either updated or pushed into core TB as I absolutely cannot work without using VI and have downgraded to 60 as well

**Comment by sej7278** [comment]
yes whilst you can customize the from address in TB 68 its a pain when you reply to an email with a different from address to that which you received it using as you forgot to change it manually.

**Comment by mcanonic** [comment]
Ok, it seems to me that we have not any chance to see this add-on up and running on TB. My question is: is there any place in the thunderbird website, where a new feature can be asked? If all of us (and, hopefully, many others) push up this request, maybe we will have some chance to see it again available in TB. What do you think?

**Comment by St3fanBC** [comment]
I'm also ready to donate if that's what it takes. This add-on is a core reason for me using TB in the first place.

**Comment by bl00dy1837** [comment]
+1 ready to donate.

**Comment by miskovics** [comment]
+1 ready to donate.

**Comment by canihavesomecoffee** [comment]
From what I see in the linked Thunderbird issue, there are patches to implement certain functionality in the core, but discussion is ongoing on how it should actually behave and how it should be enabled/disabled...

**Comment by mcanonic** [comment]
Could you send us the link of the discussion, please?
--
Sent from Libero Mail for Android Wednesday, 12 February 2020, 06:41PM +01:00 from Willem  notifications@github.com :

**Comment by canihavesomecoffee** [comment]
It was mentioned earlier in the issue by absorb-it, but here it is again: https://bugzilla.mozilla.org/show_bug.cgi?id=1518025

**Comment by alexolog** [comment]
Willing to donate for TB68 support

**Comment by absorb-it** [comment]
Hi, thanks for all the offers for donation - it's really great to see that you are willing to support this feature. On the other hand, the implementation process is not dependent on donations and some working implementation was already proposed in the mentioned thunderbird ticket. But there seems to be no intention to let this feature be part of thunderbird. Sorry to say that there is nothing which can be done from our side. I accept the decision from thunderbird-people to not add every feature - but I'm still unsure about which mail-client can be a suitable replacement in the future.

Thanks again for all your support, Rene

**Comment by alexolog** [comment]
What required functionality is missing in the new TB?

**Comment by github-normalnutzer** [comment]
While answering/replying, choose automatically the "TO:" Address as Sender Address, 
if no Address is found in the TB identities.

**Comment by alexolog** [comment]
> While answering/replying, choose automatically the "TO:" Address as Sender Address,
> if no Address is found in the TB identities.

This functionality was missing from earlier versions of TB and yet VI was able to implement it.
Obviously the question is what functionality required to implement the above is missing in TB68.

**Comment by KIMWW** [comment]
> some working implementation was already proposed in the mentioned thunderbird ticket. But there seems to be no intention to let this feature be part of thunderbird. 
> 
> Thanks again for all your support, Rene

Thank you Rene to try to solve this issue.
What can we do to make Thunderbird people to change their mind?? Would it help if all of us [post a message HERE?](https://bugzilla.mozilla.org/show_bug.cgi?id=1518025)

**Comment by absorb-it** [comment]
> What can we do to make Thunderbird people to change their mind?? Would it help if all of us [post a message HERE?](https://bugzilla.mozilla.org/show_bug.cgi?id=1518025)

I have not the slightest idea if this helps, some people tried but without success. If the intention is missing to drive thunderbird that way, we can't change that.

best regards, Rene

**Comment by absorb-it** [comment]
Hi all, good news, the related modification was confirmed by thunderbirds code maintainer. Some changes are still required, which will take some additional time - I'm busy right now but I will surely implement this during the next weeks/months. Most important: the path is free for the implementation! Yeah! Thanks for all your support, Regards, Rene

**Comment by github-normalnutzer** [comment]
Make my Day. :-)
This is really nice.
Great.....

**Comment by KIMWW** [comment]
> related modification was confirmed by thunderbirds code maintainer. [...] I'm busy right now but I 
> will surely implement this during the next weeks/months. [...] Rene

Excellent! Donations that has been offered will help. 
It's amazing what thoughts power can do.

**Comment by chris-fp** [comment]
This is fantastic news and I look forward to upgrading my TB to a newer version!

I'd still like to suggest a donation for all your work and effort over the years though.

**Comment by AlexVallat** [comment]
Plenty of comments about donations, but I don't actually see anywhere to donate to support this effort. Am I missing something obvious?

**Comment by absorb-it** [comment]
Hi, there is no donation required from my side, I am interested in this feature anyway and I like to give back something for all the free software I am using in my day-to-day live - which a lot of people create without any payment.
But thanks for all the offers, it really showed me how important this feature is for you. And as already said, without your support in this case we could not have convinced the thunderbird people how important this feature is for us.

Feel free to donate to some other software project which might really require this money to free some time for their development efforts. Or donate this money to some humanitarian organisation which engages to help people around the world, for instance https://sea-watch.org/ which is one project I'm engaged with.

Thanks again, I probably will need some time to finalize these patches and than it will take a while till we will get some release which has this feature included. But we don't have to look for another email client anymore :)

Best regards, Rene

**Comment by chconnor** [comment]
Donation made to sea-watch!

Thanks so much for your hard work on this plugin through the years, and for getting it in to core. I'm thrilled to hear there will be some kind of functionality coming, and I'm following the bugzilla thread eagerly.

:-)

**Comment by c33s** [comment]
@absorb-it i haven't digged through the mozilla thread and through the patch code, maybe you are so kind and can discribe the the new featureset here.

will the new implementation in the core be as safe as the current plugin? for me it's important that i simply send ("from" email) with the email i got the mail (previous "to" email) from the exact same account the message got delivered to. also it's quite important to see the (auto)-choosen email i will send from before i press the send button.
the confirmation dialog of your plugin often saved me from accidentaly send from the wrong acount/email.

do i miss something or does bugzilla require you to attach patches and making code review by pasting code segments with change requests as comment instead of having a github/gitlab like pull/merge-request feature with clean code comments? :facepalm:

**Comment by absorb-it** [comment]
@chconnor, thanks for the donation...

Hi @c33s, the new thunderbird featureset won't give you all the features you had with the extension, but because of the new webex-plugin schema the extension won't work like before. With the patch [1294027](https://bugzilla.mozilla.org/show_bug.cgi?id=1294027) learned to send mails with some flexible senders address, with the current patch [1518025](https://bugzilla.mozilla.org/show_bug.cgi?id=1518025) thunderbird will be able to answer with any email which can be found in some dedicated headers ("to", "cc", "envelope-to, x-original-to"). The list of headers can be changed with some (hidden) option.

Now if this is achieved, some main functionality will be in core thunderbird. And it will be robust over all upcoming changes of thunderbird, which is great. You will see a warning once the email is modified and the email will be sent with exactly the address mentioned in the from field, but there is no warning in the moment of sending anymore.

The final step to come close to yesterdays extension will be to add some option to store the used email address together with the recipient probably in the addressbook. I will have a look at this later, it's unclear if this can or better will be done by some small extension or if this should be part of the core code as well, we will see.

> do i miss something or does bugzilla require you to attach patches...

Yes, that's the way it is. But it works, the biggest problem is to build thunderbird locally and rebuild this once the underlaying mozilla changed, or some tests changed or whatever. There are options for that doing this online as well, but non of this is optimized for limited/expensive bandwidth, and that's what I usually have. Anyway, it's close to be done, yeah!

Best regards, Rene

**Comment by jfriend00** [comment]
Count me as another user anxiously waiting for the new add-on.  I was a long time user of virtual identify with Postbox until they dropped support for all add-ons while back.  I dealt with it by manually creating separate identifies for every "from" email I was trying to use, but that got to be quite a pain. Now, I'm back to Thunderbird and desperately in need of an ability to arbitrarily set the from address for any email I'm sending.  Like some others, I use a single domain and a different mailbox name in front of the domain for every website I interact with and thus need that ability to set the From address to whatever I want.  I like that virtual identify did some of this automatically, but I'd even accept 100% manual control (have to manually edit the "from" address on every email if that was all there was.

**Comment by timkgh** [comment]
@jfriend00 You can use an older version of TB. Not ideal from a security perspective, but VI works well on TB 52.9.1

**Comment by jfriend00** [comment]
@timkgh I'm aware of that from the commentary here, but I don't intend to install the older version.  I am anxiously waiting to get back to using the virtual identity I started with quite a few years ago.  All I really need is just the ability to type in my own custom "from address" so I can make an email be from whichever of my mailboxes I want.  I don't need all the other features it had (though some are useful time savers).

**Comment by GrahamCobb** [comment]
> @jfriend00 You can use an older version of TB. Not ideal from a security perspective, but VI works well on TB 52.9.1

Also works with TB 60.9.0. I have held that version on my Debian Testing system - which will, no doubt, cause a problem sometime.

**Comment by absorb-it** [comment]
Hi, this option to edit your senders address manually is already in core thunderbird and can be accessed by the identity dropdown menu with the neest thunderbird. There must be something like "Edit From Address" or the like in the identity selection dropdown menu  - this only can be seen even on 60.9.1 if Virtual Identity Extension is not installed. Regards, Rene

**Comment by jfriend00** [comment]
@absorb-it - Sure enough.  I had an "Identity Chooser" add-on that was blocking access to that feature.  Thanks.  I will appreciate the better features of Virtual Identity if you get it working again, but at least I'm up an running with "Edit From Address" for now.  Thx.

**Comment by JoelAzaria** [comment]
Rene,
Thank you so much for your dedication to this project after so many months of struggle.   I hope this core implementation will allow to reply from whatever email address in the 'To:' field, even if this is not a configured mailbox in TB.   

I hand out many bespoke email addresses with companies (makes filtering and spam control easier.) This normally works great till I forget to change the reply address then BAM! mass confusion because I reply from a strange address.
The key is I use catch-all or cpanel forwarder address for this so these are not actual existing mailboxes.  Hopefully the implementation will simply read whatever is in the "To:" field and send based on that.

Thank you for all your hard work and dedication to this project.  
I hope I can gift sea-watch in your name.

**Comment by markfilipak** [comment]
Hi! I stumbled onto this thread and I'm glad I did. I'm stuck with TBird 52.7.0 running in a Linux virtual machine. VI is version 0.9.36,

I've encountered a couple of bugs in VI, but of course that doesn't matter anymore. Before abandoning VI, I thought I'd seek the help of the good folks here. What's your advice?

Regards, and Good Health!
Mark Filipak.

**Comment by cgraefe** [comment]
Hi Rene,
This is probably the wrong place but I just realized that I could actually install Thunderbird 78 now without destroying my personal workflow. And after reading the [Bugzilla Report](https://bugzilla.mozilla.org/show_bug.cgi?id=1518025) I couldn't resist the the urge to send you a HUGE "thank you" for all your work to get -- against all odds -- the essence of Virtual Identity into Thunderbird core. I'm _so_ relieved that we have a solution to this problem. Thanks again for this and of course for your previous work on Virtual Identity!
Christian

**Comment by alexolog** [comment]
I am confused by the above comment.

Does VI work on TB78?

**Comment by cgraefe** [comment]
@alexolog Sorry to cause confusion. No, VI does not work on TB78 not will it work on any later version, to my knowledge. But @absorb-it has integrated (essential) parts of it directly into Thunderbird itself.

**Comment by alexolog** [comment]
I see.  Could you comment on how it works compared to VI?

Further questions to @absorb-it:

From my understanding of [1518025](https://bugzilla.mozilla.org/show_bug.cgi?id=1518025), it only works with given domain names, so if my email provider supports multiple domain will  I need to create a catch-all for each?

Also, it says that the list of headers can be customized with the `mail.compose.catchAllHeaders` preference.  How will TB decide which one to use if several of the specified headers are present and conflicting?  Is it by order?

> The final step to come close to yesterdays extension will be to add some option to store the used email address together with the recipient probably in the addressbook. I will have a look at this later, it's unclear if this can or better will be done by some small extension or if this should be part of the core code as well, we will see.

That would be a very helpful feature!  Maybe specifying a custom field in the vCard so it can synchronized via CardDAV?

Thanks again for your work!

**Comment by alexolog** [comment]
Also, it mentions `mail.compose.catchAllHeaders` as well as `mail.compose.multiEmailSupportHeaders`.  Which one is used?

**Comment by StephenMilner** [comment]
Hi. @absorb-it or anyone else who can assist.

Have these changes been taken into 78.3.3 or would I still need to patch Thunderbird to enable this?

If they have been integrated I must be doing something wrong?

**Comment by absorb-it** [comment]
Hi, sorry all of you, I really missed that discussion...

You are right, TB 78 includes the new implementation which might make the extension not required anymore. So, VI does not work on TB78+, but you can do a lot of things Virtual Identity brought to you natively now.

To get the "Virtual Identity"-feeling you have to activate the "Reply from this identity when delivery headers match:" option in your account settings. Enter your domains there, like "*@example.com, *@example.de". All mails you receive now which match these domains will be answered with the right email address now (with all other settings like signature etc. of the identity where you activated this feature)

This will not work for composing new emails (even if you already wrote some email to the recipient with some other email address), because there is no chance for TB to know which email address you used last time, except looking into some headers of some received email. This is still annoying, I know.

Speaking of the headers of receiving emails, they are checked via the "mail.compose.catchAllHeaders" preference value. The default setting is pref("mail.compose.catchAllHeaders", "envelope-to, x-original-to, to, cc");

If you reply to a list, they are used in the order mentioned. On a standard non-list reply "to" and "cc" are preferred in front of the list (if this turns out to be causing some issues, some bug should be filed to get this fixed). So for usual replys it's "to, cc, envelope-to, x-original-to, to, cc" (yeah :(

If you have this order of headers, they are all matched against any existing identity and this one is used if it matches. If there are no matches, all identities with CatchAll enabled ("reply ...  when ... headers match") are checked again, looping trough all headers-emails of the received email and checking against all catchAll-Domains added to this identity. As below:

      loop trough headers of  mail.compose.catchAllHeaders {
        loop trough all identities with catchAll enabled {
          loop trough fields in catchAll texfield (like for instance *@example.com) {
            if textfield matches email in header:
              if catchAll was with *@, use email found in header (this is virtualIdentity)
              (if catchAll has no *@ use the email of the identity)
            }
          }
        }

Hope this is somehow usable. I'm not really happy that this option is hidden behind some other feature but at least Thunderbird is usable for me (and probably you) again. This implementation is the result of some compromise, but seems to work for me. But I still think it will be really hard to document this feature and name it the right way, but at least it's available.

I will try to be a little bit faster on this channel if you have more questions :) Best regards,
Rene

**Comment by mcanonic** [comment]
Hi, maybe I'm doing something wrong.
I've added a *@uniupo.it in the account settings in order to have always my uniupo account to send email to my colleagues, but for it seems not working. I've noticed also a weird behaviour: sometimes with reply button, I get an email compose windows with a sender and "to:" with the same address but the sender is not me.
How much I miss VI...

**Comment by danfos** [comment]
I appreciate the work done to integrate VI functionality into Thunderbird core.

It works for me, also for email accounts with aliases/auto-forwarding. 

For aliases/auto-forwarding accounts you can add an additional identity using _Manage Identities_ under Account Settings. Then set the _Reply from this identity when delivery headers match_ for this secondary identity.

As for new mail, as far as I remember, VI did pop up a window to chose the account when the Send button was hit and a rule was matching. Would that still be an option?

Update: Found a good work-around in the [Identity Chooser add-on](https://github.com/speedball2001/identitychooser-mx). That forces you, before you create a new email, to select the identity/email where the mail will be sent from. It works also for forwarding email.

**Comment by zic-github** [comment]
Looks working almost good with TB78 but isn't there a way to sort of auto-register VI when composing emails?
Unlike good old VI, TB78 won't save manually edited From address, I guess.
We have to go to Accounts setting to manually add a VI in prior to use it.
Otherwise that manually edited From address will be lost.

**Comment by chconnor** [comment]
@zic-github -- you might be interested in these bug reports, especially the first one:

https://bugzilla.mozilla.org/show_bug.cgi?id=1679100
https://bugzilla.mozilla.org/show_bug.cgi?id=1679103
https://bugzilla.mozilla.org/show_bug.cgi?id=1677329
https://bugzilla.mozilla.org/show_bug.cgi?id=1677368

I don't get the impression that there is a lot of momentum on these right now, and I join you (and I suspect many others) in hoping that some hero developer is motivated to work a little more on it. It seems like an important final piece of the puzzle to really make this functional.

**Comment by mcanonic** [comment]
Yes, we hope so!

M

**Comment by timkgh** [comment]
@absorb-it What is the last known TB version where the old VI addon fully works? I've been on TB 52.9.1 with VI 0.9.36 for a long time but not sure whether there are newer versions where VI still **fully** works. IIRC there were some versions where it only worked partially.

**Comment by jab42** [comment]
hi,
i have an 60.9.1 (64-Bit) at debian where i can not see any non working parts.
jay


Am 16.11.21 um 16:08 schrieb timkgh:
> @absorb-it What is the last known TB version where the old VI addon fully works? I've been on TB 52.9.1 with VI 0.9.36 for a long time but not sure whether there are newer versions where VI still **fully** works. IIRC there were some versions where it only worked partially.
>

**Comment by martin3000** [comment]
Is there a thunderbird addon which generates sender addresses with timestamps?
So when sending mails, my From: address becomes "name202205071045@domain.com" and for every new mail I send, a new From: address is generated.

**Comment by KIMWW** [comment]
[Thunderbird 102](https://blog.thunderbird.net/2022/05/7-great-new-features-coming-to-thunderbird-102/) has been released so if you want to leave a comment at the end of the article it's time now.

**Comment by silversword411** [comment]
I've been testing 102. It looks like you have to have the identity existing, and it'll do right. 

Now need to pre-create all the identities I have (hundreds)...and figure out how to get manually typed ones auto-added as a new identity.

Looks like they're saved in `prefs.js`...but you're not supposed to edit that (and there's apparently another `user.js` file, but there's almost no docs about it). And if you create an identity, manually delete it's keys in `prefs.js` there's still a leftover pointer somewhere in identities. Haven't found that yet. 

I wonder if there's a setting/extension that will let manually entered addresses in from dropdown be automatically created as an identity.

**Comment by danfos** [comment]
> Now need to pre-create all the identities I have (hundreds)...

Yes, I am also using a different email address for each shop or party I communicate with and then the numbers go up fast.

Needing to have a separate identity for each is still the major drawback with what is now available for Thunderbird. For now I did create the identity when I needed it but whole concept of using an TB identity for this feel likes overhead. 

If I reply to an email addressed to abc@def.gh I simply want (by default) that the from address of that reply is abc@def.gh without adding that as an TB identity.

**Comment by dicer** [comment]
@silversword411 @danfos Replying to emails works without setting up new identities. Just put `*@domain.tld` into the following field on your account settings page:
![Screenshot_20220807_001757](https://user-images.githubusercontent.com/219524/183267707-1aa1e16b-498a-4811-b7a3-79bed1eb3659.png)
When sending "fresh" emails, you do need to manually edit the from address though. The VI addon remembered which recipient which address was for, which was a huge help. Wish that would be added to Thunderbird. Other than that it actually works as expected.

**Comment by martin3000** [comment]
If I send a new mail, I do "customize from address" by clicking into the "from" field. I change the adress to myname+receivername@tld.com without creating a new identity. Replying then works.

**Comment by danfos** [comment]
Yes, what @martin3000 write works also good for me but it is a manual extra step that I sometimes forget, so I prefer that have it automatic.

The settings that @dicer indicates are part of the setting for a particular identity but do not solve the problem for a scenario that I and I think @silversword411 are using. 

With "Reply from this identity when delivery headers match" you can make that emails received from given domains/addresses and I think also emails address to a certain email address use the give identity.

This is fine but not needed as the identity is already chosen as the email is addressed to the email of the identity and in that case the identity is used for the reply without this special setting.

The problem that you need one identity for each email you are replying, which is fine for numbers up to ~ 10 but not if you use a different email address for almost every shop or party you communicate with.

**Comment by martin3000** [comment]
I have ONE identity but use many many "from:" email-adresses. You don't need many identities!

**Comment by alexolog** [comment]
> @silversword411 @danfos Replying to emails works without setting up new identities. Just put `*@domain.tld` into the following field on your account settings page:
> ![Screenshot_20220807_001757](https://user-images.githubusercontent.com/219524/183267707-1aa1e16b-498a-4811-b7a3-79bed1eb3659.png)

Which "delivery headers" are checked to choose the correct identity?  VI allowed me to specify the one(s) I wanted to use.

**Comment by chconnor** [comment]
See this comment above: https://github.com/absorb-it/Virtual-Identity/issues/22#issuecomment-712382568

Recent commenters may also be interested in the bugs listed here: https://github.com/absorb-it/Virtual-Identity/issues/22#issuecomment-861013276

**Comment by KIMWW** [comment]
> See this comment above: [#22 (comment)](https://github.com/absorb-it/Virtual-Identity/issues/22#issuecomment-712382568)
> 
> Recent commenters may also be interested in the bugs listed here: [#22 (comment)](https://github.com/absorb-it/Virtual-Identity/issues/22#issuecomment-861013276)

The first bug (#1518025) has been fixed.

**Comment by silversword411** [comment]
Just a note for others, if you're trying to figure out why the wildcard setting above isn't working for you....disable your "Correct identity" and "Identity Chooser" plugins ;)

**Comment by danfos** [comment]
> I have ONE identity but use many many "from:" email-adresses. You don't need many identities!

@martin3000: Any guidance on how to add an email-address without an identity?

Will these email-address without identity also work without filling in something like that [Reply from this identity ...](https://github.com/absorb-it/Virtual-Identity/issues/22#issuecomment-1207291532)?

**Comment by chconnor** [comment]
> The first bug (#1518025) has been fixed.

Thanks -- but 1518025 is not in the list of bugs from the comment I referenced. Or I'm confused about what you meant?

**Comment by martin3000** [comment]
> @martin3000: Any guidance on how to add an email-address without an identity?

I click on "Write", then an empty mail opens. I click on my email address and a list of identities appears. At the bottom, I click "customize address" and then I can change the "from:" address to anything I like.

**Comment by alexolog** [comment]
> Just a note for others, if you're trying to figure out why the wildcard setting above isn't working for you....disable your "Correct identity" and "Identity Chooser" plugins ;)

There is a [discussion](https://github.com/dennisverspuij/tb-correctidentity) on the Correct Identity github that led to some changes.  Maybe participating in it could lead to further improvements.

**Comment by danfos** [comment]
> There is a [discussion](https://github.com/dennisverspuij/tb-correctidentity) on the Correct Identity github that led to some changes. Maybe participating in it could lead to further improvements.

I presume you mean https://github.com/dennisverspuij/tb-correctidentity/issues/21

There you write:

> What I meant was that I would like to reply from the same address that the email was originally sent to, even if it ended up being delivered to a different one.

That is exactly what I want but without having an identity for that.

That can be accomplished using "customize address" as @martin3000 pointed out, but would I like to have it automatically, again, without having an identity for that email-address.

**Comment by zic-github** [comment]
Virtual Identity saves used From address.
Now I think we need to manually add that From address in
>Account Settings > Manage Identities ... > [Add]

Isn't there any way to automatically saves the from address entered via "Customize From Address...* on reply?

**Comment by Globe51** [comment]
Is this add-on now totally dead? From reading the comments it looks like new versions of Thunderbird have some, but not all, of the features from VI, but without the an option to autocomplete when creating a new email it means I will still be stuck on Thunderbird 60.9 :(

**Comment by chconnor** [comment]
AFAIK the idea is to get the remaining functionality into TB, so I don't think this plugin is being developed, no. I haven't seen any movement on the outstanding features that are still missing in TB, so I think we need a hero developer or three to come along and make it happen (our traditional hero developer was Rene / @absorb-it, but I haven't seen a post from them in a while, so I'm not sure if they are still involved in these features.)

To those of us that use these features, it's amazing that it isn't standard in all email programs, but I guess we're the weird ones. :-)

**Comment by markfilipak** [comment]
I've found a work-around: For each of my favorite recipients, I struggle to craft a 'desired' email message (which I'm usually able to do), then I use those 'desired' email messages as templates for actual messages via "Edit As New Message". That allows me to bypass all the tinkering to get 'identities' right. In fact, I don't have to use 'identities' at all.

The Tbird developers just seem clueless regarding how an email program should work. They've crowbarred what should be simple sender-recipient pairs based on email headers into their rigid 'account' architecture, and then try to use identities to 'overcome' most of the architecture. I won't even mention what a pain it is to try to use one account's SMTP for a differing account.

PS: I'm running Tbird 102.4.2

**Comment by absorb-it** [comment]
Hi,

> Is this add-on now totally dead?

definitely! There will be no such extension again, at least from my side. Still thinking about adding something like an auto-save feature for the senders address to the collected addresses which can be re-used for new emails. But can't say when or if this happens - there needs to be time for this, you can see updates probably in the ticket https://bugzilla.mozilla.org/show_bug.cgi?id=1679100
Problem is, that with the current situation it's really hard to do this as some add-on (will require thunderbird code changes as well), therefore I can only do and maintain this if it will go into the core thunderbird code - and this is obviously also difficult and more complex. But, maybe...

best regards, Rene

**Comment by cboling2** [comment]
Rene, I just want to say "thanks" for a very useful extension and all the work you've put into making the issues it sought to solve less painful.  I'm still running TB 60 on my main PC, for this extension and one other; I have 102 on my laptop, where it's only occasionally used "in the field" for urgent business.

**Comment by githubbapoopa** [comment]
I second that.
TB 60.9.0 (portable) with Virtual Identity 0.10.3 still working perfectly since years and years here.
Couple that with spamgourmet and choose a mail provider with an open mail relay and you can also reply with your spamgourmet (or in fact any) email address >Minimal spam in your main account.

You saved me so much trouble and time.

Thanks so much!!!

**Comment by sjau** [comment]
Unfortunately this does not seem to work anymore. I did set

[x] Reply from this identity when delivery headers match: *@domain.tld

but when I actually reply to an email, it seems to chose the `Delivered-To` address over the `To` address header. The delivered-to address header is the actual email box and not the alias being used to differentiate all the services.

**Comment by larena1** [comment]
> Unfortunately this does not seem to work anymore. I did set
> 
> [x] Reply from this identity when delivery headers match: *@domain.tld
> 
> but when I actually reply to an email, it seems to chose the `Delivered-To` address over the `To` address header. The delivered-to address header is the actual email box and not the alias being used to differentiate all the services.

I can confirm that it's not working anymore on the latest update of Thunderbird :/ it used to work for like the past 2 years or so but not anymore...

**Comment by sjau** [comment]
> I can confirm that it's not working anymore on the latest update of Thunderbird :/ it used to work for like the past 2 years or so but not anymore...

The ReplyAsOriginalRecipient AddOn (from TB Addons directly or https://github.com/ThierryBZH/ReplyAsOriginalRecipient ) makes it work again.

**Comment by larena1** [comment]
Thanks but that addon is from 3 years ago and back then there wasn't even this functionality implemented in Thunderbird?

I'd rather prefer them to fix the regression they introduced recently but somebody would've to file a bug report with them and they would need to look into it...

**Comment by sjau** [comment]
> Thanks but that addon is from 3 years ago and back then there wasn't even this functionality implemented in Thunderbird?

It still works.

> I'd rather prefer them to fix the regression they introduced recently but somebody would've to file a bug report with them and they would need to look into it...

Me too, but until it's fixed I find that addon much more convenient than always copy'n'pasting the "from" address myself.

**Comment by absorb-it** [comment]
> Unfortunately this does not seem to work anymore. I did set
> 
> [x] Reply from this identity when delivery headers match: *@domain.tld
> 
> but when I actually reply to an email, it seems to chose the `Delivered-To` address over the `To` address header. The delivered-to address header is the actual email box and not the alias being used to differentiate all the services.

Which version of TB you are using? I'm using TB 115.5.2 and it does seem to work.

Best regards, Rene

**Comment by sjau** [comment]
115.6.0 (64-bit)

Really loved the original Virtual Identity Addon :)

I create alias for every serivce that I need to supply an email address - mostly in the form of   domain.tld_YYYY-MM-DD_UNIXTIMESTAMP@mydomain.tld

**Comment by larena1** [comment]
> 115.6.0 (64-bit)

I'm also on that version and can confirm it's not working for me either.

**Comment by mcanonic** [comment]
Could you guys explain me how  I can install this add-on since I've the 115.6.0 (64-bit) version and I would give it a try.

I've downloaded the zip file from github but it did not work: TB complains that the file is corrupted.

**Comment by sjau** [comment]
> ive

Search the TB addons page for `ReplyAsOriginalRecipient`

**Comment by mcanonic** [comment]
Thanks sjau.
I would just report that by clicking on the name, the link seems broken:
https://addons.thunderbird.net/en-US/thunderbird/addon/replyasoriginalrecipientup/?src=search

**Comment by sjau** [comment]
I used the addon search within Thunderbird and that worked just fine.

**Comment by KIMWW** [comment]
I'm on v. 115.6.0 an it still works for me.

**Comment by larena1** [comment]
https://github.com/mozilla/releases-comm-central/commit/aca3f4a814ad805d686034c175395866181e9293

Found the culprit but somebody should file an issue and report this to Mozilla for them to fix it

**Comment by alexolog** [comment]
I don't understand, what is the issue?
`mail.compose.catchAllHeaders` is user-changeable, is it not?

**Comment by larena1** [comment]
It is but that commit still introduced a regression that should be fixed upstream

**Comment by alexolog** [comment]
No argument from me, but Mozilla sometimes take their sweet time fixing issues, so if changing that setting can be a workaround, why not document it here and save people some grief?

**Comment by sjau** [comment]
my current workaround is using that addon for the time being.

**Comment by WolfgangDpunkt** [comment]
After a long search, the solution was so easy! Great.
I tried all the identity addons, no one works as well as this natively built-in solution. But I only found it by chance, please put it as reference to the Github homepage of this project. :-)

> To get the "Virtual Identity"-feeling you have to activate the "Reply from this identity when delivery headers match:" option in your account settings. Enter your domains there, like "*@example.com, *@example.de". All mails you receive now which match these domains will be answered with the right email address now (with all other settings like signature etc. of the identity where you activated this feature)

**Comment by absorb-it** [comment]
Hi Wolfgang, good point. Just added this text to the homepage as suggested.

Best regards, Rene

**Comment by c33s** [comment]
@WolfgangDpunkt i have to oppose your comment, the absolute best identity solution is the one from rene.

**Comment by mcanonic** [comment]
I hear that cloude is very good to create add on extension for TB. Does anyone tied to ask to Cloude to update this addon?

**Comment by c33s** [comment]
@mcanonic the problem is not about updating the addon but that the plugin api of thunderbird changed and simply does not support the featureset of this plugin anymore.

i sticked with the old thunderbird version as long as i clould. now after trying tb140 i really want back. i don't know if it's possible but maybe putting the old thunderbird in a flatpak and replacing the tls lib is possible (according ai it is but well ai...)

@betterbird looks interesting, maybe they can be convinced they integrate this addon.

**Comment by KIMWW** [comment]
> [@mcanonic](https://github.com/mcanonic) the problem is not about updating the addon but that the plugin api of thunderbird changed and simply does not support the featureset of this plugin anymore.
> 
> i sticked with the old thunderbird version as long as i clould. now after trying tb140 i really want back. i don't know if it's possible but maybe putting the old thunderbird in a flatpak and replacing the tls lib is possible (according ai it is but well ai...)

Get Thunderbird ESR

**Comment by alexolog** [comment]
I am using Betterbird with a catch-all identity enabled, and `mail.compose.catchAllHeaders` set to `X-Delivered-to, envelope-to, x-original-to, to, cc`

**Comment by mcanonic** [comment]
Does anyone experience with this addon:
https://services.addons.thunderbird.net/EN-gb/thunderbird/addon/correct-identity/

called correct identity?
Thanks
M

**Comment by Betterbird** [comment]
> [@Betterbird](https://github.com/Betterbird) looks interesting, maybe they can be convinced they integrate this addon.

What should be integrated? The add-on itself states (https://github.com/absorb-it/Virtual-Identity/blob/ng_0.9/README.md):

_This is an old extension, the main feature is now part of thunderbird base code. You don't need this extsnion (sic) anymore._

**Comment by KIMWW** [comment]
> Does anyone experience with this addon:
> https://services.addons.thunderbird.net/EN-gb/thunderbird/addon/correct-identity/
This add-on can:

    Have full, unrestricted access to Thunderbird, and your computer

**Comment by mcanonic** [comment]
ok, sorry for bothering you. Last question concerning the feature already provided by TB.
I have a work account and a personal account. In the work account, I have set "reply from this identity when delivery headers match: *@workdomain.it

Then I tried from my personal email account inside TB to send an mail to some colleague (that is, with @workdomain.it) but it didn't work. My colleague receives an email from my personal email address and TB did not warn me.

Am I wrong?

Edit: does this feature work only for reply and not for email composing?

**Comment by Betterbird** [comment]
> ```
> Have full, unrestricted access to Thunderbird, and your computer
> ```

That (threatening text) is standard for Webextension Experiments which really have access to everything.

> Edit: does this feature work only for reply and not for email composing?

There are two things odd here. You configure something for account W and are surprised when is doesn't work for account P?
And yes, as the UI suggests: _**Reply** from this identity when delivery headers match._

**Comment by mcanonic** [comment]
Sorry about that, but that is exactly what I miss. Whit this old extension, I was able to send email using the correct sender.
If by mistake I was going to send a work email from my personal email, VI automatically replaces the personal email with the work email.

**Comment by chris-fp** [comment]
I'm still using tb 61 with VI as I've not seen anything in TB or BB that matches the functionality I need.

I don't use the + functionality, I use [companyname/website_FQDN]@testdomain.com  so for example for an account with the bbc, the email I use with them is [bbc@testdomain.com](mailto:bbc@testdomain.com) or disney+ is [disneyplus@testdomain.com](mailto:disneyplus@testdomain.com) and when I reply I need to have the from address to be matched.   Using identities looks insane as I'd need hundreds of them.  TB61+VI just simply handles this automatically.

Am I missing something or is this possible using TB/BB now?

**Comment by EkriirkE** [comment]
> Am I missing something or is this possible using TB/BB now?

Nothing automatic.  Currently what is "working" for me is 1 identity per host, and I use the "Customize From Address" option and edit the email address for the selected domain each time

**Comment by realityexists** [comment]
> when I reply I need to have the from address to be matched
> Am I missing something or is this possible using TB/BB now?

I have the exact same use case as you. This part is mostly possible with TB/BB now, though it might not detect your email address (bbc@testdomain.com) as well as VI did: https://bugzilla.mozilla.org/show_bug.cgi?id=1679103

When you enable "Reply from this identity when delivery headers match:" in account settings, and configure it to match *@testdomain.com, and hit reply on an email sent to bbc@testdomain.com it will set the From email to bbc@testdomain.com (with a yellow warning bar saying that it did not match any email address exactly).

The big parts of VI still missing are:

- it sets only the From *email address*, not the name. So even if your identity normally sends email from `Chris <bbc@testdomain.com>` the "catch-all" functionality in TB will set the From field to just `bbc@testdomain.com`: https://bugzilla.mozilla.org/show_bug.cgi?id=1677329
- when you compose a new email you have to manually set the from address (though you can edit that in the email compose dialog itself) - TB will not remember the address you previously used for that recipient: https://bugzilla.mozilla.org/show_bug.cgi?id=1679100

**Comment by mcanonic** [comment]
I'm working on an addon to replace the VI, now it is under review by TB guys.
If you want try it, as a beta tester,  I can send you the xpi file but please keep in mind that it is a "working in progress" project. I've been using my addon for 3 days now and so far so good

If the TB guys approve my addon, I'll let you know. Usually it takes 7-10 days after submission, so I think that in the middle of next week I'll have an answer for you.

**Comment by chconnor** [comment]
Great news, @mcanonic, thanks for working on it. I don't know if I'll have time to test it but I will certainly try to make time. 

Should we assume that it works for both new emails and response emails, as VI did?

Did you create this with Claude?

**Comment by absorb-it** [comment]
Hi @mcanonic ,

> I'm working on an addon to replace the VI, now it is under review by TB guys.

thanks for working on that addon. Can you please add some link to the source code or the beta xpi to separate the comments on your work from this ticket.

Regards, Rene

**Comment by mcanonic** [comment]
> Great news, [@mcanonic](https://github.com/mcanonic), thanks for working on it. I don't know if I'll have time to test it but I will certainly try to make time.
> 
> Should we assume that it works for both new emails and response emails, as VI did?

It works for new email for sure and I didn't follow well the problem with the reply but if you provide an example I'll try to reproduce it.

> 
> Did you create this with Claude?

Yes.

**Comment by mcanonic** [comment]
> Hi [@mcanonic](https://github.com/mcanonic) ,
> 
> > I'm working on an addon to replace the VI, now it is under review by TB guys.
> 
> thanks for working on that addon. Can you please add some link to the source code or the beta xpi to separate the comments on your work from this ticket.

Did you mean to create a "new issue" here?

Btw, the code right now is here:
https://www.dropbox.com/scl/fi/u7agc7ajuejc3vavee7p5/smart-sender.xpi?rlkey=6y2er6fzg66tar9f5se7gz3co&st=5bxrj9bz&dl=0

Once again, it is under review by TB guys. I'm using it, it seems ok at least for me, but since the email is a crucial resource (anyone has some good AI agent to manage the email?) for all of us, consider yourself as a beta tester if you try to install the xpi.

M

> 
> Regards, Rene

**Comment by absorb-it** [comment]
> > Hi [@mcanonic](https://github.com/mcanonic) ,
> > > I'm working on an addon to replace the VI, now it is under review by TB guys.
> > 
> > 
> > thanks for working on that addon. Can you please add some link to the source code or the beta xpi to separate the comments on your work from this ticket.
> 
> Did you mean to create a "new issue" here?

No, just find your place (github, gitlab, whatever) to add your add-on and track (and later fix) the related issues. It's ok to have some link to your project here, but this is a ticket referring to the old Virtual Identity extension. Thanks for your understanding,

Rene

**Comment by jab42** [comment]
> Btw, the code right now is here:

Should i post comments to the Addon here, or is there another place meanwhile?

**Comment by mcanonic** [comment]
Here:
https://github.com/mcanonic/smart-sender

In the xpi I added a switch to temporarily pause the addon.

**Comment by jab42** [comment]
> Here:

Saw my 2 issues there?

**Comment by mcanonic** [comment]
Yes, I'll reply there.
